analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -159,15 +159,18 @@ private static IEnumerable<SecondaryLocation> CheckConstructor(TypeDeclarationSy
         private static bool ImplementsISerializable(ITypeSymbol typeSymbol) =>
             typeSymbol != null
             && typeSymbol.IsPubliclyAccessible()
-            && typeSymbol.AllInterfaces.Any(IsOrImplementsISerializable);
+            && typeSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable)
+            && OptsInForSerialization(typeSymbol);
+
+        private static bool OptsInForSerialization(ITypeSymbol typeSymbol) =>
+            typeSymbol is INamedTypeSymbol namedTypeSymbol
```

Comment:
I would move this check to `ImplementsISerializable` and change the declaration to receive a INamedTypeSymbol instead of an IType Symbol. This way `OptsInForSerialization` would be just a sequence of OR, instead of A && (B || C || D)